File: dotnet/Logify.Alert.Core/Tests/Logger/JsonTextWriterLoggerTests.cs
#if DEBUG
using System;
using System.IO;
using System.Text;
using NUnit.Framework;

namespace DevExpress.Logify.Core.Internal.Tests {
    [TestFixture]
    public class JsonTextWriterLoggerTests {
        TextWriter writer;
        StringBuilder content;
        ILogger logger;

        [SetUp]
        public void Setup() {
            this.content = new StringBuilder();
            this.writer = new StringWriter(content);
            this.logger = new TextWriterLogger(writer);
        }
        [TearDown]
        public void TearDown() {
            this.content = null;
            this.writer.Dispose();
            this.logger = null;
        }
        [Test]
        public void BeginWriteObject() {
            logger.BeginWriteObject("testObject");
            Assert.AreEqual(string.Empty, content.ToString());
        }
        [Test]
        public void BeginWriteObjectNoName() {
            logger.BeginWriteObject(String.Empty);
            Assert.AreEqual(string.Empty, content.ToString());
        }
        [Test]
        public void EndWriteObject() {
            logger.EndWriteObject("testObject");
            Assert.AreEqual(string.Empty, content.ToString());
        }
        [Test]
        public void WriteValueString() {
            logger.BeginWriteObject(String.Empty);
            logger.WriteValue("variable", "value");
            logger.EndWriteObject(String.Empty);
            Assert.AreEqual("{\r\n\"variable\":\"value\"\r\n}\r\n", content.ToString());
        }
        [Test]
        public void WriteValueStringWithDotInName() {
            logger.BeginWriteObject(String.Empty);
            logger.WriteValue("devexpress.com", "value");
            logger.EndWriteObject(String.Empty);
            Assert.AreEqual("{\r\n\"devexpress\uff0Ecom\":\"value\"\r\n}\r\n", content.ToString());
        }
        [Test]
        public void WriteValueStringSpecialCharacters() {
            logger.BeginWriteObject(String.Empty);
            logger.WriteValue("variable", "special: \r\n\u0012\t\b\f\"\\;");
            logger.EndWriteObject(String.Empty);
            Assert.AreEqual("{\r\n" + @"""variable"":""special: \r\n\u0012\t\b\f\""\\;""" + "\r\n}\r\n", content.ToString());
        }
        [Test]
        public void WriteValueBool() {
            logger.BeginWriteObject(String.Empty);
            logger.WriteValue("variable", true);
            logger.EndWriteObject(String.Empty);
            Assert.AreEqual("{\r\n\"variable\":true\r\n}\r\n", content.ToString());
        }
        [Test]
        public void BeginWriteArray() {
            logger.BeginWriteArray("testArray");
            Assert.AreEqual(String.Empty, content.ToString());
        }
        [Test]
        public void BeginWriteArrayNoName() {
            logger.BeginWriteArray(String.Empty);
            Assert.AreEqual(String.Empty, content.ToString());
        }
        [Test]
        public void EndWriteArray() {
            logger.EndWriteArray("testArray");
            Assert.AreEqual(String.Empty, content.ToString());
        }

        [Test]
        public void WriteValueArray() {
            int[] testArray = {1, 2, 3};
            logger.BeginWriteObject(String.Empty);
            logger.WriteValue("testArray", testArray);
            logger.EndWriteObject(String.Empty);
            Assert.AreEqual("{\r\n\"testArray\":[\"1\",\"2\",\"3\"]\r\n}\r\n", content.ToString());
        }

        [Test]
        public void WriteValueEmptyArray() {
            int[] testArray = {};
            logger.WriteValue("testArray", testArray);
            Assert.AreEqual("", content.ToString());
        }
    }
}
#endif

File: java/Core/src/test/java/com/devexpress/logify/alert/core/logger/JsonTextWriterLoggerTests.java
package com.devexpress.logify.alert.core.logger;

import com.devexpress.logify.alert.core.collectors.CollectorTestsBase;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class JsonTextWriterLoggerTests extends CollectorTestsBase {

    @Before
    public void setUp() {
        super.setUp();
    }

    @After
    public void tearDown() throws IOException {
        super.tearDown();
    }

    @Test
    public void testWriteObject() {
        logger.beginWriteObject("testObject");
        logger.endWriteObject("testObject");

        JsonElement element = parseJsonReport(true).get("testObject");
        assertNotNull(element);
        assertTrue(element.isJsonObject());
        assertTrue(element.getAsJsonObject().entrySet().isEmpty());
    }

    @Test
    public void testWriteObjectNoName() {
        logger.beginWriteObject("");
        logger.endWriteObject("");

        JsonElement element = parseJsonReport(false);
        assertTrue(element.isJsonObject());
        assertTrue(element.getAsJsonObject().entrySet().isEmpty());
    }

    @Test
    public void testWriteValueString() {
        logger.writeValue("variable", "value");

        JsonElement element = parseJsonReport(true).get("variable");
        assertNotNull(element);
        assertTrue(element.isJsonPrimitive());
        assertEquals("value", element.getAsString());
    }

    @Test
    public void testWriteValueStringSpecialCharacters() {
        String text = " \"\\ \u0001 \u0010\" \n\n\n abc";
        logger.writeValue("variable", text);

        JsonElement element = parseJsonReport(true).get("variable");
        assertNotNull(element);
        assertEquals(text, element.getAsString());
    }

    private void testWriteBoolean(boolean value) {
        logger.writeValue("variable", value);

        JsonElement element = parseJsonReport(true).get("variable");
        assertNotNull(element);
        assertTrue(element.isJsonPrimitive());
        assertEquals(value, element.getAsBoolean());
    }

    @Test
    public void testWriteValueBooleanTrue() {
        testWriteBoolean(true);
    }

    @Test
    public void testWriteValueBooleanFalse() {
        testWriteBoolean(false);
    }

    @Test
    public void testWriteInt() {
        logger.writeValue("variable", 42);

        JsonElement element = parseJsonReport(true).get("variable");
        assertNotNull(element);
        assertTrue(element.isJsonPrimitive());
        assertEquals(42, element.getAsInt());
    }

    @Test
    public void testWriteList() {
        List<Object> list = new ArrayList<Object>();
        list.add(4);
        list.add("test");
        list.add(true);
        logger.writeValue("variable", list);

        JsonElement element = parseJsonReport(true).get("variable");
        assertNotNull(element);
        assertTrue(element.isJsonArray());

        JsonArray array = element.getAsJsonArray();
        assertEquals(list.size(), array.size());
        for (int i = 0; i < list.size(); i++) {
            assertEquals(list.get(i).toString(), array.get(i).getAsString());
        }
    }

    @Test
    public void testWriteArray() {
        logger.beginWriteArray("testArray");
        logger.endWriteArray("testArray");

        JsonElement element = parseJsonReport(true).get("testArray");
        assertNotNull(element);
        assertTrue(element.isJsonArray());
        assertEquals(0, element.getAsJsonArray().size());
    }

    @Test
    public void testWriteMultipleElements() {
        logger.writeValue("key1", 42);
        logger.writeValue("key2", "text");
        logger.writeValue("key3", false);

        logger.beginWriteObject("object");
        logger.writeValue("key4", true);
        logger.endWriteObject("object");

        logger.beginWriteArray("array");
        logger.beginWriteObject("");
        logger.endWriteObject("");
        logger.beginWriteObject("");
        logger.endWriteObject("");
        logger.endWriteArray("array");

        new JsonParser().parse("{" + logger.toString() + "}");
    }

}
